base/dict.jl

```diff
@@ -901,14 +901,31 @@ struct PersistentDict{K,V} <: AbstractDict{K,V}
     @noinline function KeyValue.set(::Type{PersistentDict{K, V}}, ::Nothing, key, val) where {K, V}
         new{K, V}(HAMT.HAMT{K, V}(key => val))
     end
-    @noinline function KeyValue.set(dict::PersistentDict{K, V}, key, val) where {K, V}
+    @noinline @Base.assume_effects :effect_free function KeyValue.set(dict::PersistentDict{K, V}, key, val) where {K, V}
         trie = dict.trie
         h = HAMT.HashState(key)
         found, present, trie, i, bi, top, hs = HAMT.path(trie, key, h, #=persistent=# true)
         HAMT.insert!(found, present, trie, i, bi, hs, val)
         return new{K, V}(top)
     end
-    @noinline function KeyValue.set(dict::PersistentDict{K, V}, key) where {K, V}
+    @noinline @Base.assume_effects :nothrow :effect_free function KeyValue.set(dict::PersistentDict{K, V}, key::K, val::V) where {K, V}
```

Comment:
Upon further investigation, it appears that modifying this single line is enough to remove the call to the dead `PersistentDict` constructor. But, this seems to stem from an inference bug, where we're deleting statements without actually proving `terminates`. 
```julia
julia> @noinline Base.@assume_effects :effect_free :nothrow function foo(n)
           s = 0
           while true
               if n - rand(1:10) > 0
                   s += 1
               else
                   break
               end
           end
           return s
       end^C

julia> code_typed((Int,)) do n
           foo(n)
           nothing
       end
1-element Vector{Any}:
 CodeInfo(
1 ─     return Main.nothing
) => Nothing
```
So, I'd like to address this bug first and then revisit this PR.